feat: 회원 CRUD 기능 (#12)

green4469 opened the pull request.

green4469 committed:
feat: 사용자 회원가입 폼을 위한 컴포넌트
생성

green4469 committed:
feat: 사용자 회원 가입 폼

green4469 committed:
feat: 사용자 엔티티

green4469 committed:
feat: 사용자 레포지토리

green4469 committed:
feat: 사용자 회원가입 서비스, dto

green4469 committed:
feat: 사용자 데이터베이스 저장 실패 예외

green4469 committed:
feat: 사용자 회원가입 컨트롤러

green4469 committed:
feat: 사용자 정보 조회

green4469 committed:
feat: 사용자 정보 수정

ep1stas1s commented on src/frontend/src/components/user/SignUpForm.vue:
```diff
@@ -0,0 +1,83 @@
+      />
+
+      <div class="text-center">
+        <v-btn @click="requestSignUp" block tile color="grey" dark>회원 가입</v-btn>
```
한 줄에 속성 하나!

ep1stas1s commented on src/frontend/src/components/user/SignUpForm.vue:
```diff
@@ -0,0 +1,83 @@
+    },
+    methods: {
+      requestSignUp: function () {
+        const request = require('request');
```
이 부분은 정확히 모르는데, 공통으로 사용하는 건 뺄 수 있는 방법을 찾아보는 게 좋을 것 같아요
사장님에게 여쭤보거나ㅎㅎ

ep1stas1s commented on src/frontend/src/components/user/UserProfile.vue:
```diff
@@ -0,0 +1,59 @@
+      request.get({
+        url: `${window.location.origin}/api/users/${this.userId}`,
+      }, function (error, response, body) {
+        window.console.log('statusCode:', response && response.statusCode); // Print the response status code if a response was received
```
log는 test용인가용?

yunheehyeon replied:
네 지금은 지우겠습니다.

ep1stas1s commented on src/frontend/src/components/user/SignUpForm.vue:
```diff
@@ -0,0 +1,83 @@
+
+      <v-text-field
+        v-model="signUpFormData.password"
+        :append-icon="showPassword ? 'mdi-eye' : 'mdi-eye-off'"
```
삼항연산자... 그는 과연 javascript인가? HTML인가? (backtick vs single quote)

Deocksoo replied:
저는 backtick에 한표!

ep1stas1s commented on src/main/java/woowahan/anifarm/tecolearning/user/controller/UserController.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @GetMapping("/{userId}")
+    public ResponseEntity<UserInfoDto> read(@PathVariable long userId) {
```
```suggestion
    public ResponseEntity<UserInfoDto> read(@PathVariable Long userId) {
```
Entity의 경우 wrapper type을 사용하는데 이 부분은 어떡하는 것이 좋을까요?

yunheehyeon replied:
Entity와 Dto에서 wrapper 사용 나머지는 기본 타입으로 사용합시다!

ep1stas1s commented on src/main/java/woowahan/anifarm/tecolearning/user/controller/UserController.java:
```diff
@@ -0,0 +1,35 @@
+
+    @PutMapping
+    public ResponseEntity update(@RequestBody UserUpdateDto userUpdateDto) {
+        userService.update(userUpdateDto, 1);
```
`1`은 test용인가용?

yunheehyeon replied:
네 아직 로그인이 구현되지 않아서...ㅎ

ep1stas1s commented on src/main/java/woowahan/anifarm/tecolearning/user/domain/User.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public void update(User updateUser) {
+        this.nickName = updateUser.getNickName();
```
```suggestion
        this.nickName = updateUser.nickName;
```

yunheehyeon replied:
왜 바꿔야한다고 생각하시나요? 고래의 생각을 알려주세요

ep1stas1s commented on src/main/java/woowahan/anifarm/tecolearning/user/service/UserService.java:
```diff
@@ -0,0 +1,42 @@
+        this.userRepository = userRepository;
+    }
+
+    public long save(UserCreateDto userCreateDto) {
```
Wrapper type?

yunheehyeon replied:
위와 동일합니다.

ep1stas1s commented on src/main/java/woowahan/anifarm/tecolearning/user/service/UserService.java:
```diff
@@ -0,0 +1,42 @@
+        return UserInfoDto.from(findById(userId));
+    }
+
+    public User findById(long userId) {
```
다른 곳에서 사용할까봐 public으로 남겨두신건가요?

yunheehyeon replied:
지금은 아니지만 다른 서비스에서 사용할 수 있을거 같아서 public으로 했습니다.

green4469 replied:
StudyService 에서 persist 된 User (즉, id 를 갖는 유저) 가 필요할 수 있을거라 생각했습니다. (예를 들면 Study 에 List<User> 로 스터디원을 가지고 있고 여기에 새 유저를 id 로 추가하고 싶은 상황)

ep1stas1s commented on src/main/java/woowahan/anifarm/tecolearning/user/service/UserService.java:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    public UserInfoDto findInfoDtoById(long userId) {
```
흠...

yunheehyeon replied:
흠... user는 중복되어서 지웠습니다.  dto로 변환 할 때 assembler 클래스를 사용논의 필요?

ep1stas1s commented on src/test/java/woowahan/anifarm/tecolearning/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,77 @@
+                .body(Mono.just(params), Map.class)
+                .exchange()
+                .expectStatus()
+                .isOk();
```
```
webTestClient.post()
                .uri("/api/users/signup")
                .body(Mono.just(new UserCreateDto("green4469@woowahan.com", "moomindev", "")), UserCreateDto.class)
                .exchange()
                .expectStatus()
                .isOk();
```
Body로 UserCreateDto 객체를 보내는 건 어떨까요?

green4469 replied:
Mono.just 로 안하고 바로 보내는건가요?

ep1stas1s reviewed (changes requested):
첫 주차부터 고생하셨습니다!🦄🦛
짧은 시간안에 vue부터 back까지...
승차감 좋은 게 벌써 느껴지네요~😎

악플다는 게 생각보다 재밌네요^^

ep1stas1s commented on src/frontend/src/components/user/SignUpForm.vue:
```diff
@@ -0,0 +1,83 @@
+        name="input-10-1"
+        label="패스워드 입력"
+        hint="최소 8 글자 이상 입력하세요."
+        counter
```
잘 못 이해하는 것일 수도 있지만,
counter같은 경우는 닉네임에 `:counter="10"`같이 쓰는 것도 좋을 것 같아요!

yunheehyeon replied:
counter 속성을 사용한 이유는 사용자가 작성한 문자의 길이를 표시하기 위해서 사용했습니다.
counter="10" 을 사용하면 어떤 이점이 있을 까요?

Deocksoo commented on src/frontend/src/components/user/SignUpForm.vue:
```diff
@@ -0,0 +1,83 @@
+      }
+    },
+    methods: {
+      requestSignUp: function () {
```
아마 requestSignUp() 으로 축약 선언이 가능할거에요, 한번 확인부탁드립니다 :)

Deocksoo commented on src/frontend/src/components/user/UserUpdateForm.vue:
```diff
@@ -0,0 +1,75 @@
+        label="이메일"
+        readonly
+        outlined
+      />
```
요 form 부분의 중복을 제거할 수 있을지도 한번 생각해 봐야겠네요!

Deocksoo commented on src/frontend/src/components/user/UserUpdateForm.vue:
```diff
@@ -0,0 +1,75 @@
+          body: { nickName: this.nickName },
+          json: true,
+        }, function (error, response, body) {
+          window.console.log('error:', error); // Print the error if one occurred
```
주석

ddu0422 commented on src/main/java/woowahan/anifarm/tecolearning/user/controller/UserController.java:
```diff
@@ -0,0 +1,35 @@
+@RestController
+@RequestMapping("/api/users")
+public class UserController {
+    private UserService userService;
```
final 키워드를 사용하여 불변을 유지하는건 어떨까요?

ddu0422 commented on src/main/java/woowahan/anifarm/tecolearning/user/domain/User.java:
```diff
@@ -0,0 +1,38 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false)
```
length를 지정하는건 어떨까요?

yunheehyeon replied:
기획의 문제라고 생각하는데 논의가 필요한거 같습니다.

ddu0422 commented on src/main/java/woowahan/anifarm/tecolearning/user/exception/UserCreateException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class UserCreateException extends RuntimeException {
+    public UserCreateException(Throwable cause) {
+        super("사용자 생성 실패", cause);
```
상수로 빼는건 어떨까요?

ddu0422 commented on src/main/java/woowahan/anifarm/tecolearning/user/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+
+public class UserNotFoundException extends RuntimeException {
+
+    public static final String ERROR_MSG = "사용자를 찾지 못했습니다.";
```
convention !!
해당 예외를 테스트하는 코드를 추가하는건 어떤가요?
Exception 안에 있는 상수이니 상수명은 MESSAGE는 어떤가요?

ddu0422 commented on src/main/java/woowahan/anifarm/tecolearning/user/controller/UserController.java:
```diff
@@ -0,0 +1,35 @@
+    @PostMapping("/signup")
+    public ResponseEntity create(@RequestBody UserCreateDto userCreateDto) {
+        userService.save(userCreateDto);
+        return ResponseEntity.ok().build();
```
convention !!

green4469 replied:
"/signup" url 네이밍 말씀하시는 건가요?

ddu0422 commented on src/frontend/src/components/user/SignUpForm.vue:
```diff
@@ -0,0 +1,83 @@
+</template>
+
+<script>
+  export default {
```
`use strict`를 작성해주세요 !

ddu0422 commented on src/frontend/src/components/user/UserProfile.vue:
```diff
@@ -0,0 +1,59 @@
+      userId: Number,
+    },
+    mounted() {
+      let temp = this;
```
temp가 변수 타입으로 const를 가져도 property의 값은 바꿀 수 있지 않나요?

ddu0422 commented on src/main/java/woowahan/anifarm/tecolearning/user/domain/User.java:
```diff
@@ -0,0 +1,38 @@
+@NoArgsConstructor(access = AccessLevel.PUBLIC)
+@Getter
+public class User {
+    @Id
```
convention !!

ddu0422 commented on src/main/java/woowahan/anifarm/tecolearning/user/service/UserService.java:
```diff
@@ -0,0 +1,42 @@
+
+@Service
+public class UserService {
+    private UserRepository userRepository;
```
여기도 final 키워드를 명시하여 불변을 유지해보아요~

ddu0422 commented on src/main/java/woowahan/anifarm/tecolearning/user/service/UserService.java:
```diff
@@ -0,0 +1,42 @@
+        try {
+            return userRepository.save(userCreateDto.toEntity()).getId();
+        } catch (Exception error) {
+            throw new UserCreateException(error);
```
log를 추가하는 건 어떨까요?

yunheehyeon replied:
언제 로그를 사용할 지 정해보면 어떨까요?

ddu0422 commented on src/test/java/woowahan/anifarm/tecolearning/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,33 @@
+@ExtendWith(MockitoExtension.class)
+class UserServiceTest {
+    @InjectMocks
+    private UserService userService;
```
convention을 참고해볼까요?

ddu0422 commented on src/test/java/woowahan/anifarm/tecolearning/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,77 @@
+        params.put("nickName", "moomin");
+
+        webTestClient.post()
+                .uri("/api/users/signup")
```
/api/users/ 가 공통이네요
Controller에서 중복을 제거 한 것처럼 여기도 중복을 제거하는건 어떨까요?

green4469 replied:
상수처리해서 중복을 제거할 수는 있겠지만, 가독성이 떨어진다고 생각합니다.

예시)
``` java
webTestClient.post()
    .uri("/api/users/signup")

webTestClient.post()
    .uri(USERS_API_URL + "/signup")
```

후자의 경우엔 가독성이 더 떨어지지 않나요?

ddu0422 commented on src/frontend/src/components/user/SignUpForm.vue:
```diff
@@ -0,0 +1,83 @@
+        label="패스워드 입력"
+        hint="최소 8 글자 이상 입력하세요."
+        counter
+        @click:append="showPassword = !showPassword"
```
@click은 vue 속성이 아닐까요?

ddu0422 reviewed (commented):
고생하셨어요~ 
간단한 피드백 남겼습니다

green4469 committed:
refactor: HTML 태그 속성은 한 줄에 하나만 나열한다.

green4469 committed:
refactor: 개발용 로그 삭제

뷰는 함수의 마지막 인자를 사용하지 않으면 컴파일 에러를 발생시킨다. 마지막 인자를 사용하지 않을 경우엔
사용할 인자까지만 남기고 나머지는 없앤다. 또한 코드 블락 내에 아무 코드도 존재하지 않으면 컴파일 에러를 발생시킨다.
이 경우엔 주석 한 줄만 추가해도 컴파일이 가능하므로 TODO 메시지를 주석으로 남겼다.

green4469 committed:
refactor: 멤버 변수는 변하지 않을 경우 final 로 선언한다.

green4469 committed:
refactor: 에러 메시지 상수화, 상수명은 MESSAGE

green4469 committed:
refactor: 엄격한 자바스크립트 문법을 사용한다

green4469 committed:
refactor: 재 할당되지 않는 변수는 const 로 선언한다.

green4469 committed:
refactor: 엔티티 객체의 애노테이션이 붙어있는 필드는 위 아래로 한 칸씩 띈다.

green4469 committed:
refactor: Mock 객체를 주입받는 객체는 inject 를 prefix 로 붙인다.

green4469 committed:
refactor: vue 속성을 상위에 위치한다

green4469 committed:
feat: 사용자 삭제 요청시 비활성화

green4469 committed:
refactor: 컨트롤러 테스트 코드에 중복된 로직을 추상화

Deocksoo reviewed (commented):
확인 부탁드려요!

green4469 committed:
refactor: DbUnit으로 테스트 데이터 초기화

green4469 committed:
refactor: 같은 타입의 다른 인스턴스의 필드 사용시 getter말고 직접 접근

Deocksoo commented on src/main/java/woowahan/anifarm/tecolearning/user/domain/AccountStatus.java:
```diff
@@ -0,0 +1,6 @@
+package woowahan.anifarm.tecolearning.user.domain;
+
+public enum AccountStatus {
```
👍  amdin이라던지 차차 여기에 추가하면 관리하기 편하겠네요
